Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -271,7 +270,16 @@ class ArraySortedMap : public SortedMapBase {
    *     requested key.
    */
   const_iterator lower_bound(const K& key) const {
-    return std::lower_bound(begin(), end(), key, key_comparator_);
+    const_iterator iter = begin();
+    const_iterator iter_end = end();
+
+    for (; iter != iter_end; ++iter) {
+      auto cmp = comparator_.Compare(iter->first, key);
+      if (cmp != util::ComparisonResult::Ascending) {
+        return iter;
+      }
+    }
+    return iter_end;
```

Comment:
std::lower_bound is a binary search, i.e. logn, but this is O(n). I think you could still use std::lower_bound:

```
return std::lower_bound(begin(), end(), key, [&](pair<K,V>& a, b) {
  return util::Ascending(comparator_.Compare(a.first, b.first));
});
```

(again, untested)

Reply:
Both cases are really O(1) because n is never greater than 15. The effect that's most likely to dominate is cache locality, which is likely to benefit linear search more than binary. 

However, I'm unwilling to construct a benchmark for this or supply alternate versions based on item size so I'll just defer to your (almost correct) suggestion.

Reply:
Right; I had forgotten about the fact that these are always small. No benchmark required by me if you want to switch back... but I still like using lower_bound as it's the more obvious thing to do here.